ocaml/xapi/message_forwarding.ml

```diff
@@ -6429,4 +6437,88 @@ functor
               ~base_build_id ~base_version ~base_release ~to_version ~to_release
         )
     end
+
+    module Observer = struct
+      let create ~__context ~name_label ~name_description ~hosts ~attributes
+          ~endpoints ~components ~enabled =
+        info "Observer.create: name_label=%s" name_label ;
+        Local.Observer.create ~__context ~name_label ~name_description ~hosts
+          ~attributes ~endpoints ~components ~enabled
+
+      let register ~__context ~self ~host =
+        info "Observer.register: self=%s" (observer_uuid ~__context self) ;
+        let local_fn = Local.Observer.register ~self ~host in
+        let client_fn session_id rpc =
+          Client.Observer.register ~rpc ~session_id ~self ~host
+        in
+        do_op_on ~__context ~host ~local_fn client_fn
+
+      let unregister ~__context ~self ~host =
+        info "Observer.unregister: self=%s" (observer_uuid ~__context self) ;
+        let local_fn = Local.Observer.unregister ~self ~host in
+        let client_fn session_id rpc =
+          Client.Observer.unregister ~rpc ~session_id ~self ~host
+        in
+        do_op_on ~__context ~host ~local_fn client_fn
+
+      let destroy ~__context ~self =
+        info "Observer.destroy: self=%s" (observer_uuid ~__context self) ;
+        Local.Observer.destroy ~__context ~self
+
+      let set_hosts ~__context ~self ~value =
+        let uuid = observer_uuid ~__context self in
+        info "Observer.set_hosts: self=%s value=%s" uuid
+          (String.concat ";" (List.map (host_uuid ~__context) value)) ;
+        Local.Observer.set_hosts ~__context ~self ~value
```

Comment:
I think this should call some unregister and register an hosts affected, other the hosts field will become out of sync with the hosts that are actually gathering data for the observer

Reply:
That is what set hosts does

Reply:
A more verbose explanation. Set_hosts is responsible for setting the new DB state and also for calculating hosts that need to have it registered and those unregistered. It then calls these functions on that, I am doing that here within this function. Same as below